File: src/main/java/org/cubexmc/metro/integration/MapGeometry.java
package org.cubexmc.metro.integration;

import org.bukkit.Location;
import org.cubexmc.metro.model.RoutePoint;
import org.cubexmc.metro.model.Stop;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

final class MapGeometry {

    private static final double SAME_COORDINATE_EPSILON = 0.0001D;

    private MapGeometry() {
    }

    static List<RoutePoint> orthogonalRoutePoints(List<RoutePoint> routePoints, String worldName) {
        List<RoutePoint> result = new ArrayList<>();
        if (routePoints == null || worldName == null || worldName.isBlank()) {
            return result;
        }

        RoutePoint previous = null;
        for (RoutePoint current : routePoints) {
            if (current == null || !worldName.equals(current.worldName())) {
                continue;
            }
            if (previous == null) {
                addIfDifferent(result, current);
                previous = current;
                continue;
            }

            double dx = Math.abs(current.x() - previous.x());
            double dz = Math.abs(current.z() - previous.z());
            if (dx > SAME_COORDINATE_EPSILON && dz > SAME_COORDINATE_EPSILON) {
                RoutePoint elbow = dx >= dz
                        ? new RoutePoint(worldName, current.x(), previous.y(), previous.z())
                        : new RoutePoint(worldName, previous.x(), previous.y(), current.z());
                addIfDifferent(result, elbow);
            }
            addIfDifferent(result, current);
            previous = current;
        }
        return result;
    }

    static Optional<StopBounds> stopBounds(Stop stop) {
        if (stop == null || stop.getCorner1() == null || stop.getCorner2() == null) {
            return Optional.empty();
        }

        Location corner1 = stop.getCorner1();
        Location corner2 = stop.getCorner2();
        if (corner1.getWorld() == null || corner2.getWorld() == null || !corner1.getWorld().equals(corner2.getWorld())) {
            return Optional.empty();
        }

        double minX = Math.min(corner1.getBlockX(), corner2.getBlockX());
        double maxX = Math.max(corner1.getBlockX(), corner2.getBlockX()) + 1.0D;
        double minY = Math.min(corner1.getBlockY(), corner2.getBlockY());
        double maxY = Math.max(corner1.getBlockY(), corner2.getBlockY()) + 1.0D;
        double minZ = Math.min(corner1.getBlockZ(), corner2.getBlockZ());
        double maxZ = Math.max(corner1.getBlockZ(), corner2.getBlockZ()) + 1.0D;
        return Optional.of(new StopBounds(corner1.getWorld().getName(), minX, minY, minZ, maxX, maxY, maxZ));
    }

    private static void addIfDifferent(List<RoutePoint> points, RoutePoint point) {
        if (points.isEmpty() || !samePosition(points.get(points.size() - 1), point)) {
            points.add(point);
        }
    }

    private static boolean samePosition(RoutePoint left, RoutePoint right) {
        return left.worldName().equals(right.worldName())
                && Math.abs(left.x() - right.x()) <= SAME_COORDINATE_EPSILON
                && Math.abs(left.y() - right.y()) <= SAME_COORDINATE_EPSILON
                && Math.abs(left.z() - right.z()) <= SAME_COORDINATE_EPSILON;
    }

    record StopBounds(String worldName, double minX, double minY, double minZ, double maxX, double maxY, double maxZ) {
        double centerX() {
            return (minX + maxX) / 2.0D;
        }

        double centerY() {
            return (minY + maxY) / 2.0D;
        }

        double centerZ() {
            return (minZ + maxZ) / 2.0D;
        }
    }
}

File: src/main/java/org/cubexmc/metro/integration/MapGeometry.kt
package org.cubexmc.metro.integration

import java.util.Optional
import org.cubexmc.metro.model.RoutePoint
import org.cubexmc.metro.model.Stop

class MapGeometry private constructor() {
    data class StopBounds(
        private val worldName: String,
        private val minX: Double,
        private val minY: Double,
        private val minZ: Double,
        private val maxX: Double,
        private val maxY: Double,
        private val maxZ: Double,
    ) {
        fun worldName(): String = worldName

        fun minX(): Double = minX

        fun minY(): Double = minY

        fun minZ(): Double = minZ

        fun maxX(): Double = maxX

        fun maxY(): Double = maxY

        fun maxZ(): Double = maxZ

        fun centerX(): Double = (minX + maxX) / 2.0

        fun centerY(): Double = (minY + maxY) / 2.0

        fun centerZ(): Double = (minZ + maxZ) / 2.0
    }

    companion object {
        private const val SAME_COORDINATE_EPSILON = 0.0001

        @JvmStatic
        fun orthogonalRoutePoints(routePoints: List<RoutePoint>?, worldName: String?): List<RoutePoint> {
            val result = ArrayList<RoutePoint>()
            if (routePoints == null || worldName == null || worldName.isBlank()) {
                return result
            }

            var previous: RoutePoint? = null
            for (current in routePoints) {
                if (current == null || worldName != current.worldName()) {
                    continue
                }
                if (previous == null) {
                    addIfDifferent(result, current)
                    previous = current
                    continue
                }

                val last = previous
                val dx = kotlin.math.abs(current.x() - last.x())
                val dz = kotlin.math.abs(current.z() - last.z())
                if (dx > SAME_COORDINATE_EPSILON && dz > SAME_COORDINATE_EPSILON) {
                    val elbow = if (dx >= dz) {
                        RoutePoint(worldName, current.x(), last.y(), last.z())
                    } else {
                        RoutePoint(worldName, last.x(), last.y(), current.z())
                    }
                    addIfDifferent(result, elbow)
                }
                addIfDifferent(result, current)
                previous = current
            }
            return result
        }

        @JvmStatic
        fun stopBounds(stop: Stop?): Optional<StopBounds> {
            if (stop == null) {
                return Optional.empty()
            }

            val corner1 = stop.corner1 ?: return Optional.empty()
            val corner2 = stop.corner2 ?: return Optional.empty()
            val world = corner1.world ?: return Optional.empty()
            if (corner2.world == null || world != corner2.world) {
                return Optional.empty()
            }

            val minX = kotlin.math.min(corner1.blockX, corner2.blockX).toDouble()
            val maxX = kotlin.math.max(corner1.blockX, corner2.blockX) + 1.0
            val minY = kotlin.math.min(corner1.blockY, corner2.blockY).toDouble()
            val maxY = kotlin.math.max(corner1.blockY, corner2.blockY) + 1.0
            val minZ = kotlin.math.min(corner1.blockZ, corner2.blockZ).toDouble()
            val maxZ = kotlin.math.max(corner1.blockZ, corner2.blockZ) + 1.0
            return Optional.of(StopBounds(world.name, minX, minY, minZ, maxX, maxY, maxZ))
        }

        private fun addIfDifferent(points: MutableList<RoutePoint>, point: RoutePoint) {
            if (points.isEmpty() || !samePosition(points[points.size - 1], point)) {
                points.add(point)
            }
        }

        private fun samePosition(left: RoutePoint, right: RoutePoint): Boolean =
            left.worldName() == right.worldName() &&
                kotlin.math.abs(left.x() - right.x()) <= SAME_COORDINATE_EPSILON &&
                kotlin.math.abs(left.y() - right.y()) <= SAME_COORDINATE_EPSILON &&
                kotlin.math.abs(left.z() - right.z()) <= SAME_COORDINATE_EPSILON
    }
}
